src/src/PluginStructs/P104_data_struct.cpp

```diff
@@ -0,0 +1,2312 @@
+
+# include "../Helpers/ESPEasy_Storage.h"
+# include "../Helpers/Numerical.h"
+# include "../WebServer/Markup_Forms.h"
+# include "../WebServer/WebServer.h"
+# include "../WebServer/Markup.h"
+# include "../WebServer/HTML_wrappers.h"
+# include "../ESPEasyCore/ESPEasyRules.h"
+# include "../Globals/ESPEasy_time.h"
+
+# include <vector>
+# include <MD_Parola.h>
+# include <MD_MAX72xx.h>
+
+// Needed also here for PlatformIO's library finder as the .h file
+// is in a directory which is excluded in the src_filter
+
+# if defined(P104_USE_NUMERIC_DOUBLEHEIGHT_FONT) || defined(P104_USE_FULL_DOUBLEHEIGHT_FONT)
+void createHString(String& string); // Forward definition
+# endif // if defined(P104_USE_NUMERIC_DOUBLEHEIGHT_FONT) || defined(P104_USE_FULL_DOUBLEHEIGHT_FONT)
+void reverseStr(String& str);       // Forward definition
+
+/****************************************************************
+ * Constructor
+ ***************************************************************/
+P104_data_struct::P104_data_struct(MD_MAX72XX::moduleType_t _mod,
+                                   taskIndex_t              _taskIndex,
+                                   int8_t                   _cs_pin,
+                                   uint8_t                  _modulesize)
+  : mod(_mod), taskIndex(_taskIndex), cs_pin(_cs_pin), modules(_modulesize) {
+  if (Settings.InitSPI < 1) {
+    addLog(LOG_LEVEL_ERROR, F("DOTMATRIX: Required SPI not enabled. Initialization aborted!"));
+  } else {
+    P = new MD_Parola(mod, cs_pin, modules);
+  }
+}
+
+/*******************************
+ * Initializer/starter
+ ******************************/
+bool P104_data_struct::begin() {
+  if (!initialized) {
+    loadSettings();
+    initialized = true;
+  }
+
+  if ((P != nullptr) && (cs_pin > -1)) {
+    addLog(LOG_LEVEL_INFO, F("dotmatrix: P->begin() called"));
+    P->begin(expectedZones);
+    # ifdef P104_USE_BAR_GRAPH
+    pM = P->getGraphicObject();
+    # endif // ifdef P104_USE_BAR_GRAPH
+    return true;
+  }
+  return false;
+}
+
+# define P104_ZONE_SEP   '\x02'
+# define P104_FIELD_SEP  '\x01'
+# define P104_ZONE_DISP  ';'
+# define P104_FIELD_DISP ','
+
+/**************************************
+ * loadSettings
+ *************************************/
+void P104_data_struct::loadSettings() {
+  if (taskIndex < TASKS_MAX) {
+    LoadCustomTaskSettings(taskIndex, (uint8_t *)&StoredSettings, sizeof(StoredSettings.bufferSize));
+    uint16_t structDataSize = StoredSettings.bufferSize + sizeof(StoredSettings.bufferSize);
+    # ifdef P104_DEBUG_DEV
+    String log;
+
+    if (loglevelActiveFor(LOG_LEVEL_INFO) &&
+        log.reserve(54)) {
+      log  = F("P104: loadSettings stored Size: ");
+      log += structDataSize;
+      log += F(" taskindex: ");
+      log += taskIndex;
+      addLog(LOG_LEVEL_INFO, log);
+    }
+    # endif // ifdef P104_DEBUG_DEV
+    LoadCustomTaskSettings(taskIndex, (uint8_t *)&StoredSettings, structDataSize); // Read only actual data
```

Comment:
You're loading the `StoredSettings` twice, but took me a while to realize you first only read the size of the object and then the rest.
However, the allocate buffer size is limited to `P104_SETTINGS_BUFFER`, which doesn't have to be the same as the stored `StoredSettings.bufferSize`
So that's another reason to allocate this buffer as for example std::vector which you then can make 2 bytes larger to store the length also.
In the `tP104_StoredSettings` you can then have a function like this to access the buffer:
```
char& operator[](uint16_t index) { return &(buffer[index]); }
char* get() {return &(buffer[2]);}
```